src/poolgirl_internal_directory.erl

```diff
@@ -0,0 +1,66 @@
+%% @private
+%% Poolgirl - A sexy Erlang worker pool factory
+
+-module(poolgirl_internal_directory).
+
+-include_lib("stdlib/include/ms_transform.hrl").
+
+-export([new/0,
+         register/2,
+         find/2]).
+
+%%%
+%%% Types
+%%%
+
+-opaque t() :: ets:tab().
+-export_type([t/0]).
+
+%%%
+%%% Exported functions
+%%%
+
+-spec new() -> t().
+new() ->
+    ets:new('poolgirl.internal_directory', [public]).
+
+-spec register(t(), atom()) -> boolean().
+register(Directory, Name) ->
+    case find(Directory, Name) of
+        {ok, ExistingPid} ->
+            maybe_over_register(Directory, Name, ExistingPid);
+        error ->
+            ets:insert_new(Directory, {Name, self()})
+    end.
+
+-spec find(t(), atom()) -> {ok, pid()} | error.
+find(Directory, Name) ->
+    case ets:lookup(Directory, Name) of
+        [{Name, Pid}] ->
+            {ok, Pid};
+        [] ->
+            error
+    end.
+
+%%%
+%%% Local functions
+%%%
+
+maybe_over_register(Directory, Name, ExistingPid) ->
+    (ExistingPid =:= self() orelse not is_process_alive(ExistingPid))
```

Comment:
The `ExistingPid =:= self()` is to allow for supervisors to re-register when its not their intention to do so (say, upon their `:init/2` - the place from which they register - being called an n-th time during release upgrades.)